File: client/src/game/utils/asset.lua
local json = require 'src.vendor.json'

local Asset = {}

local _levelCache = {}
local _imageCache = {}
local _soundCache = {}

function Asset.getLevel(file)
  if _levelCache[file] then
    return _levelCache[file]
  else
    local result = love.filesystem.newFile(file):read()
    local level = json.decode(result)
    _levelCache[file] = level
  end

  return _levelCache[file]
end

function Asset.getImage(file)
  if _imageCache[file] then
    return _imageCache[file]
  else
    local image = love.graphics.newImage(file)
    image:setWrap('repeat', 'repeat')
    _imageCache[file] = image
  end

  return _imageCache[file]
end

function Asset.getSound(file)
  if _soundCache[file] then
    return _soundCache[file]
  else
    local sound = love.audio.newSource(file)
    _soundCache[file] = sound
  end

  return _soundCache[file]
end

return Asset

File: packages/client/src/game/utils/asset.ts
const imageCache: Map<string, Image> = new Map()
const audioCache: Map<string, Source> = new Map()

export const loadImage = (filepath: string): Image => {
  const cached = imageCache.get(filepath)

  if (cached) {
    return cached
  } else {
    const image = love.graphics.newImage(filepath)
    image.setWrap('repeat', 'repeat')

    imageCache.set(filepath, image)

    return image
  }
}

export const loadAudio = (filepath: string) => {
  if (audioCache.get(filepath)) {
    return audioCache.get(filepath)
  } else {
    const audio = love.audio.newSource(filepath, 'static')
    audioCache.set(filepath, audio)

    return audio
  }
}
